### PR TITLE
 install: add chart-checker

### PR DESCRIPTION
**Pull request by squeed**
This adds a go chart-checker that adds certain Cilium-specific checks.
 
The first linter ensures that the preflight mode can be installed alongside an existing installation.

### CODE REVIEW

File: install/chart-checker/go.sum [hunk 1]
```diff
@@ -0,0 +1,34 @@
+github.com/Masterminds/semver/v3 v3.2.0 h1:3MEsd0SM6jqZojhjLWWeBY+Kcjy9i6MQAeY7YgDP83g=
```
Comment (squeed) [inline install/chart-checker/go.sum]: oops this file can be deleted.

File: install/chart-checker/linters/linters.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	decoder := yaml.NewYAMLOrJSONDecoder(buf, 4096)
+	for {
+		u := unstructured.Unstructured{}
+		if err := decoder.Decode(&u); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, fmt.Errorf("failed to parse manifest as k8s object: %w", err)
+		}
+		if u.GetKind() == "" {
```
Comment (gandro) [inline install/chart-checker/linters/linters.go]: What is this check doing?

File: install/chart-checker/chart-linter.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (gandro) [inline install/chart-checker/chart-linter.go]: Consider adding this to the `tools/` folders instead. We have similar checks there as well.

**Review by gandro — COMMENTED** [review]
This is a really cool idea! Thanks, looks good overall. One question and one comment

File: install/chart-checker/chart-linter.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+		},
+
+		SilenceUsage: true,
+	}
+
+	flags := cmd.PersistentFlags()
+	flags.StringVar(&chartPath, "chart-path", "", "Path to helm chart")
+	cmd.Flags().StringArrayVar(&values, "set", []string{}, "Set helm values on the command line (can specify multiple or separate values with commas: key1=val1,key2=val2)")
+
+	cmd.Execute()
```
Comment (tommyp1ckles) [inline install/chart-checker/chart-linter.go]: should check the error?

File: install/chart-checker/linters/linters.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		sets = append(sets, "--set", v)
+	}
+
+	cmd := exec.Command("helm", "template",
+		"--namespace", DefaultNamespace,
+		name, chartPath,
+		"--version", PlaceholderVersion,
+	)
+	cmd.Args = append(cmd.Args, sets...)
+	log.Println(cmd.Args)
```
Comment (tommyp1ckles) [inline install/chart-checker/linters/linters.go]: nit: we kind of use a mix log and fmt Print calls?

File: install/chart-checker/linters/linters.go (hunk 2 above)
Comment (tommyp1ckles) [reply install/chart-checker/linters/linters.go]: If the object is not a k8s resource we ignore?

**Comment by tommyp1ckles** [comment]
Nice work!

File: install/chart-checker/linters/linters.go (hunk 2 above)
Comment (squeed) [reply install/chart-checker/linters/linters.go]: Sometimes the YAMLOrJSONDecoder does funny things if it encounters an empty string; this skips that (I'll add a comment)

File: install/chart-checker/linters/linters.go (hunk 5 above)
Comment (squeed) [reply install/chart-checker/linters/linters.go]: oh, good point.

**Comment by squeed** [comment]
I've picked this back up - PTAL, thanks!

**Review by gandro — APPROVED** [review]
Awesome, thanks!

**Review by tommyp1ckles — APPROVED** [review]

File: install/kubernetes/Makefile [hunk 6]
```diff
@@ -66,6 +66,7 @@ lint:
 		|| (echo -e "$$crd not found in $(CHART_FILE).\nPlease update the chart to include $$crd."; exit 1); \
 	done
 	$(QUIET)$(HELM) lint --with-subcharts --values ./cilium/values.yaml ./cilium
+	go run ../../tools/chart-checker --chart-path ./cilium
```
Comment (ti-mo) [inline install/kubernetes/Makefile:69]: Nit: when specifying the full package path, Go will run the local version of a command when run from a directory under a go.mod. This is equivalent and a little clearer:
```suggestion
	go run github.com/cilium/cilium/tools/chart-checker --chart-path ./cilium
```

**Pull request closed without merging**